packages/next/next-server/server/image-optimizer.ts

```diff
@@ -0,0 +1,230 @@
+import { IncomingMessage, ServerResponse } from 'http'
+import { join } from 'path'
+import { mediaType } from '@hapi/accept'
+import { createReadStream, promises } from 'fs'
+import { createHash } from 'crypto'
+import Server from './next-server'
+import { fileExists } from '../../lib/file-exists'
+
+let sharp: typeof import('sharp')
+//const AVIF = 'image/avif'
+const WEBP = 'image/webp'
+const PNG = 'image/png'
+const JPEG = 'image/jpeg'
+const MIME_TYPES = [/* AVIF, */ WEBP, PNG, JPEG]
+const CACHE_VERSION = 1
+
+export async function imageOptimizer(
+  server: Server,
+  req: IncomingMessage,
+  res: ServerResponse
+) {
+  const { nextConfig, distDir } = server
+  const { sizes = [], domains = [] } = nextConfig?.experimental?.images || {}
+  const { url: reqUrl = '/', headers } = req
+  const { query } = parse(reqUrl, true)
+  const { url, w, q } = query
+  const proto = headers['x-forwarded-proto'] || 'http'
+  const host = headers['x-forwarded-host'] || headers.host
+  const mimeType = mediaType(req.headers.accept, MIME_TYPES) || ''
+
+  if (!url) {
+    res.statusCode = 400
+    res.end('"url" parameter is required')
+    return { finished: true }
+  } else if (Array.isArray(url)) {
+    res.statusCode = 400
+    res.end('"url" parameter cannot be an array')
+    return { finished: true }
+  }
+
+  let absoluteUrl: URL
+  try {
+    absoluteUrl = new URL(url)
+
+    if (
+      Array.isArray(domains) &&
+      domains.length > 0 &&
+      !domains.includes(absoluteUrl.hostname)
+    ) {
+      res.statusCode = 400
+      res.end('"url" parameter is not allowed')
+      return { finished: true }
+    }
+  } catch (_error) {
+    // url was not absolute so assuming relative url
+    try {
+      absoluteUrl = new URL(url, `${proto}://${host}`)
+    } catch (__error) {
+      res.statusCode = 400
+      res.end('"url" parameter is invalid')
+      return { finished: true }
+    }
+  }
+
+  if (!w) {
+    res.statusCode = 400
+    res.end('"w" parameter (width) is required')
+    return { finished: true }
+  } else if (Array.isArray(w)) {
+    res.statusCode = 400
+    res.end('"w" parameter (width) cannot be an array')
+    return { finished: true }
+  }
+
+  if (!q) {
+    res.statusCode = 400
+    res.end('"q" parameter (quality) is required')
+    return { finished: true }
+  } else if (Array.isArray(q)) {
+    res.statusCode = 400
+    res.end('"q" parameter (quality) cannot be an array')
+    return { finished: true }
+  }
+
+  const width = parseInt(w, 10)
+
+  if (!width || isNaN(width)) {
+    res.statusCode = 400
+    res.end('"w" parameter (width) must be a number greater than 0')
+    return { finished: true }
+  }
+
+  if (Array.isArray(sizes) && sizes.length > 0 && !sizes.includes(width)) {
+    res.statusCode = 400
+    res.end(`"w" parameter (width) of ${width} is not allowed`)
+    return { finished: true }
+  }
+
+  const quality = parseInt(q)
+
+  if (isNaN(quality) || quality < 1 || quality > 100) {
```

Comment:
any reasons we don't support `q=0`?

Reply:
This is a percent and `sharp` doesn't support 0% quality. It throws with the following:

```
Error: Expected integer between 1 and 100 for quality but received 0 of type number
```